### DISCUSSION TITLE
[API Proposal]: InterpolatedStringHandler allow method overload that has the "name" of the interpolated values.

### DISCUSSION THREAD

**Discussion by LokiMidgard** [dotnet/csharplang#7537] [Language Ideas] [answered]
### Background and motivation

In some casese it may be of relevance to know from where the value was recived.

E.g.: A structual logger may want to know the variable to use when logging.

### API Proposal

```csharp

public void AppendFormatted<T> (T value, int alignment, string? format, string name);

```

name will contain the text wiritten in source code.

### API Usage

```csharp
subject = "Mail";
logger.LogInformation($"Create entity {subject}");
```

will call `public void AppendFormatted<T> (T value, int alignment, string? format, string name);` with
`T`: `string`
`value:`: `"Mail"`
`alignment`: 0
`format`: `null`
`name`: `"subject"`

### Alternative Designs

```csharp
public void AppendFormatted<T> (T value, int alignment, string? format, System.Linq.Expression name);
```

An alternateive, or additon could be to use Expressions to get more information.

### Risks

Besides an increased complexexty I currently see no risk.
I can't anticipate the performance costs if Expressiontrees are used, but for the string variant I don't expect a high cost.

**Comment by HaloFour** [comment, marked as answer]
You can use [CallerArgumentExpressionAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callerargumentexpressionattribute?view=net-7.0) to achieve this with a custom interpolated string handler:

```cs
public void AppendFormatted<T> (
    T value,
    int alignment,
    string? format,
    [CallerArgumentExpression("value")] string? name = null) { ... }
```

**Reply by HaloFour** [reply]
Note that this will contain the expression used within the interpolation hole.  If that's just a variable name, it will be that variable name, but if it's a more complicated expression, it will be that whole expression, e.g. "x + y" or "name.ToUpper()".

**Reply by LokiMidgard** [reply]
Perfect, I wasn't aware that I could just add CallerArgument stuff.

Thank you :)
